### DISCUSSION TITLE
Messing With Randomness: How To Prevent Back-to-Back Repetition?

### DISCUSSION THREAD

**Discussion by coltmay** [nss-day-cohort-47/C47-Discussions#63] [Help Me] [answered]
I'm building a small little web app that randomly displays a phrase every time a button is clicked.  Much like Dad Jokes.

But I'm building in the randomness myself, it's just something I've been wanting to work more with.

The randomness comes from this snippet of code taking the length of the array containing the phrase objects, then passing that link through:
```
// `length` is provided from array length
Math.floor((Math.random() * length))
```
It's there a way I can generate a phrase that is guaranteed not to be a direct repetition of the current phrase?  I was thinking of utilizing a do while loop somewhere in this mess*, but I've not had much success so far.

I'm essentially making a fetch to that DB every time a button is pressed, I'm fine with things repeating, but I want to page to change no matter what once a button is pressed.

What little I've researched has pointed me toward the _Fisher-Yates Algorithm_, should I look more into that?

Here's a [link](https://github.com/coltmay/calm.), if you're interested.  Bear in mind, it is nothing very special.  Just a proof of concept more than anything.

***

*My thought is to store the last output of `Math.floor((Math.random() * length))` in its own variable, then compare that variable with the new output every time it runs.  While oldOutput === newOutput, run it again.  In theory that seems like it _may_ work, right?

**Comment by coltmay** [comment, marked as answer]
Oh hey, I figured it out.

Once I ran the function responsible for the Math.random (called `randomSelector`), I just used a while loop to invoke that function again until `randomIndex` was a value that _wasn't_ `lastRandom`.

```
while (lastRandom === randomIndex) {
   randomIndex = randomSelector(length)
}
```

**Reply by BryanNilsen** [reply]
can you post the full code snippet so we can follow the flow. this sounds great!

**Reply by coltmay** [reply]
Sure.  I'm sure it's not very elegant, but it works!
```javascript
const ShowRandomPhrase = () => {
    getPhrases()
        .then(() => {
            // Set variable equal to the length of the array being fetched.
            let length = phraseLength();
            // Set variable equal to a random number between 0 and the array length.
            let randomIndex = randomSelector(length);

            // Compares newly generated random number with the last accepted random number, if identical, re-randomizes. 
            while (lastRandom === randomIndex) {
                randomIndex = randomSelector(length)
            }

            // Fetches a phrase from the phrase array, using the random number as the index.
            let randomPhrase = usePhrases()[randomIndex]

            // Converts phrase to HTML
            let phrase = DisplayPhrase(randomPhrase)

            // Injects the phrase to the DOM
            phraseElement.innerHTML = phrase;

            // Sets new value of lastRandom, for next function invoke.
            lastRandom = randomIndex;
        })
}

const randomSelector = (length) => {
    return Math.floor((Math.random() * length));
}
```

**Reply by BryanNilsen** [reply]
what happens if you only have one item (or none) in your array of phrases??

**Reply by coltmay** [reply]
I assume something goes horribly wrong...I'd better fix that later.
